--- a/arch/mips/mach-ath79/Makefile
+++ b/arch/mips/mach-ath79/Makefile
@@ -7,4 +7,5 @@ obj-y += dram.o
 obj-$(CONFIG_SOC_AR933X)	+= ar933x/
 obj-$(CONFIG_SOC_AR934X)	+= ar934x/
 obj-$(CONFIG_SOC_QCA953X)	+= qca953x/
+obj-$(CONFIG_SOC_QCA955X)       += qca955x/
 obj-$(CONFIG_SOC_QCA956X)	+= qca956x/
--- a/drivers/pinctrl/ath79/Makefile
+++ b/drivers/pinctrl/ath79/Makefile
@@ -2,3 +2,4 @@
 
 obj-$(CONFIG_PINCTRL_AR933X) += pinctrl_ar933x.o
 obj-$(CONFIG_PINCTRL_QCA953X) += pinctrl_qca953x.o
+obj-$(CONFIG_PINCTRL_QCA955X) += pinctrl_qca955x.o
