--- a/common/cmd_bootm.c
+++ b/common/cmd_bootm.c
@@ -436,7 +436,13 @@ static int bootm_start_standalone(ulong
 		setenv("filesize", buf);
 		return 0;
 	}
+	/* images.ep has been changed order in bootm_start(),
+	 * no need to change it again! */
+	#if 0
 	appl = (int (*)(int, char *[]))ntohl(images.ep);
+	#else
+	appl = (int (*)(int, char *[]))images.ep;
+	#endif
 	(*appl)(argc-1, &argv[1]);
 
 	return 0;
