--- a/common/env_sf.c
+++ b/common/env_sf.c
@@ -143,6 +143,10 @@ err_crc:
 	puts("*** Warning - bad CRC, using default environment\n\n");
 
 	set_default_env();
+#if defined(CONFIG_CMD_SAVEENV) && !defined(CONFIG_ENV_IS_NOWHERE)
+	puts ("Save the default environment\n\n");
+	saveenv();
+#endif
 }
 
 int env_init(void)
--- a/common/env_common.c
+++ b/common/env_common.c
@@ -258,6 +258,10 @@ void env_relocate (void)
 		show_boot_progress (-60);
 #endif
 		set_default_env();
+#if defined(CONFIG_CMD_SAVEENV) && !defined(CONFIG_ENV_IS_NOWHERE)
+		puts ("Save the default environment\n\n");
+		saveenv();
+#endif
 	}
 	else {
 		env_relocate_spec ();
