--- a/common/cmd_jffs2.c
+++ b/common/cmd_jffs2.c
@@ -421,6 +421,13 @@ int mtdparts_init(void)
 		part->offset = 0x00000000;
 #endif
 
+#ifdef CONFIG_DUAL_IMAGE
+		{ char *part_offset=getenv("jffs2_scan_part_offset");
+			if (part_offset != NULL) {
+				part->offset = simple_strtoul(part_offset, NULL, 0);
+			}
+		}
+#endif
 		part->sector_size = get_part_sector_size(id, part);
 
 		part->dev = current_mtd_dev;
